File: week_8_and_9/2_JavaScript/translate_to_js/my_solution.js
// U3.W8-9: Challenge you're converting


// I worked on this challenge [by myself, with: Hing Huynh].

// 2. Pseudocode



// 3. Initial Solution
var cardCheck = function cardCheck(cardNumber) {
    if (cardNumber.toString().split('').length != 16) {
        return "ERROR: Invalid number of digits"
    }
    else {
        var numberArray = cardNumber.toString().split('').map(Number);

        for (var i = 0; i < numberArray.length; i++) {
            if (i % 2 === 0) {
                numberArray[i] = numberArray[i] * 2;
            }
        }

        numberArray2 = numberArray.join('').toString().split('').map(Number);

        return eval(numberArray2.join('+')) % 10 === 0;
    }
}

// 4. Refactored Solution
var cardCheck = function cardCheck(cardNumber) {
    if (cardNumber.toString().split('').length != 16) {
        return "ERROR: Invalid number of digits"
    }
    else {
        var numberArray = cardNumber.toString().split('').map(Number);

        for (var i = 0; i < numberArray.length; i += 2) {
                numberArray[i] *= 2;
        }
        return eval(numberArray.join('').toString().split('').map(Number).join('+')) % 10 === 0;
    }
}

// 1. DRIVER TESTS/ASSERT STATEMENTS GO BELOW THIS LINE
function assert(test, message, test_number) {
  if (!test) {
    console.log(test_number + "false");
    throw "ERROR: " + message;
  }
  console.log(test_number + "true");
  return true;
}

assert(
  (cardCheck(4408041234567893) === true),
  "Card # 4408041234567893 should be valid",
  "1. "
)

assert(
  (cardCheck(4408041234567892) === false),
  "Card # 4408041234567892 should be invalid",
  "2. "
)

assert(
  (cardCheck(440804123456789) === "ERROR: Invalid number of digits"),
  "Card # 440804123456789 should give error",
  "3. "
)

cardCheck(4408041234567893);
cardCheck(4408041234567892);
cardCheck(440804123456789);





// 5. Reflection 

File: week_8_and_9/3_Ruby/fibonacci_sequence/my_solution.rb
# U3.W8-9: 


# I worked on this challenge [by myself, with: Hing Huynh].

# 2. Pseudocode



# 3. Initial Solution

def is_fibonacci?(number)
  i = 0
  arr = [0,1]
  if number == i || number == 1 
    return true
  else
    while i <= number
      i = arr[arr.length - 1] + arr[arr.length - 2]
      arr << i
    end   
  end
  arr.include?(number) ? true : false
end

# 4. Refactored Solution

def is_fibonacci?(number)
  arr = [0,1]
  true if number == 0 || number == 1
  while arr[arr.length - 1] < number
    arr << arr[arr.length - 1] + arr[arr.length - 2]
  end   
  arr.include?(number) 
end

# 1. DRIVER TESTS GO BELOW THIS LINE
def random_fibonacci
   [0,1,1,2,3,5,8,13,21,34,55,89,144,233,377,610,987,1597,2584,4181,6765,10946].sample
end

def assert(statement = "Assertion failed!")
  raise statement unless yield
end

assert { is_fibonacci?(random_fibonacci) == true }
assert { is_fibonacci?(random_fibonacci) == true }
assert { is_fibonacci?(random_fibonacci) == true }
assert { is_fibonacci?(random_fibonacci) == true }
assert { is_fibonacci?(90) == false }
assert { is_fibonacci?(0) == true }
assert { is_fibonacci?(1) == true }

# 5. Reflection 